feat(core,phrases): add GET /phrase route

IceHe opened the pull request:
<!--
  For non-English users:
  It's okay to post in your language, but remember to use English for the body (you can paste the result of Google Translate), and put everything else as attachments.
  Issues with a non-English body will be DIRECTLY CLOSED until it's updated.
-->

<!-- MANDATORY -->
## Summary
<!-- Provide detail PR description below -->
Add `GET /phrase` route

Input

- _Detected languages_ are from `Accept-Language` request header and `locale` parameter parsed by utility function `detectLanguage` ([see details](https://github.com/logto-io/logto/blob/master/packages/core/src/i18n/detect-language.ts))

Output

- _Language tag_ is returned in [`Content-Language`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Language) response header.
- _ResourceLanguage_ for i18n is returned in the response body.

Logic

- The returned language (in `Content-Language`) is selected by [design](https://www.notion.so/silverhand/AC-Localization-28dedc76c30d4a91b4f1ef6919b8b4d5#b3cb3b6f9ea6449790aaf7e72c8c1ef1).

    <img width="207" alt="image" src="https://user-images.githubusercontent.com/10594507/191914257-3cbd7f0c-6359-4422-adf0-8bf5b18c1f71.png">

- The returned language resource (in the response body):
    - If the returned language has its custom phrase, 
        - If the returned language has its built-in phrase, merge its custom and built-in phrases.
        - If the returned language has no built-in phrase, merge its custom phrase and English built-in phrase.
    - If the returned language has no custom phrase, 
        - If the returned language has its built-in phrase, merge its built-in phrase.
        - If the returned language has no built-in phrase, 
            - If English language has its custom phrase, merge English custom and built-in phrase.
            - If English language has no custom phrase, return English built-in phrase.

- Use the response guard to ensure that the translation in the final response is fully-translated.


<!-- MANDATORY -->
## Testing
<!-- How did you test this PR? -->

UTs

- GET /phrase response header `Content-Language`

![image](https://user-images.githubusercontent.com/10594507/192135201-2dd6cac9-4e2a-4b96-9697-cb6828c70182.png)

- GET /phrase process and response body

![image](https://user-images.githubusercontent.com/10594507/192135250-0d7b89b9-9845-414b-b0cb-146ec284b685.png)

- Full translation guard

![image](https://user-images.githubusercontent.com/10594507/192135345-34c1b53b-a4e7-48d6-8227-0b1068774bfb.png)

- Get phrase by supported language key

![image](https://user-images.githubusercontent.com/10594507/192135296-e494d582-1f2b-4c58-897a-33066b4746f7.png)

HTTP request

<img width="814" alt="image" src="https://user-images.githubusercontent.com/10594507/191910982-b0b2d5dc-b801-473a-93aa-bb8823c4ea15.png">

<img width="848" alt="image" src="https://user-images.githubusercontent.com/10594507/191911022-53d66974-82c5-4f41-b82a-26ae7c5206d9.png">

Locally tested.

wangsijie commented on packages/core/src/queries/custom-phrase.ts:
```diff
@@ -8,6 +8,16 @@ import { DeletionError } from '@/errors/SlonikError';
 const { table, fields } = convertToIdentifiers(CustomPhrases);
 
+export const findAllCustomLanguageKeys = async () => {
+  const rows = await envSet.pool.many<{ languageKey: string }>(sql`
```
will `many` throw an error if there is no record?

IceHe replied:
refactored and locally tested. PTAL.

```ts
export const findAllCustomLanguageKeys = async () => {
  const rows = await manyRows<{ languageKey: string }>(
    envSet.pool.query(sql`
      select ${fields.languageKey}
      from ${table}
      order by ${fields.languageKey}
    `)
  );

  return rows.map((row) => row.languageKey);
};
```

wangsijie commented on packages/core/src/routes/phrase.test.ts:
```diff
@@ -0,0 +1,290 @@
+        ...mockSignInExperience,
+        languageInfo: {
+          autoDetect: false,
+          fallbackLanguage: ptPtKey,
```
nit: can we change `ptPtKey` to something like `defaultKey`?

wangsijie commented on packages/core/src/routes/phrase.ts:
```diff
@@ -0,0 +1,131 @@
+    return getPhrase(defaultLanguageKey);
+  }
+
+  if (!hasCustomPhrase && isLanguageKeyEnum(languageKey)) {
```
What will happen if both `hasCustomPhrase` and `isLanguageKeyEnum(languageKey)` are falsy?

The if statements in this function are complicated and not easy to understand.

IceHe replied:
`isLanguageKeyEnum(languageKey)` is just used for type narrowing to pass type check when `resource[languageKey]` (`resource` is from @logto/phrases-ui)

wangsijie replied:
Then can we use `assert`?

wangsijie replied:
The current code looks like that not every situation is handled properly.

IceHe replied:
refactored. PTAL

wangsijie commented on packages/core/src/routes/phrase.ts:
```diff
@@ -0,0 +1,130 @@
+  new Error(`The final default language "${finalDefaultLanguage}" must contain a built-in phrase.`)
+);
+
+const getPhrase = async (
```
nit: we can extract this function, and test it directly, it would be more clear compare to the long route testing code

simeng-li commented on packages/core/src/routes/phrase.ts:
```diff
@@ -0,0 +1,130 @@
+    }),
+    async (ctx, next) => {
+      const interaction = await provider
+        .interactionDetails(ctx.req, ctx.res)
+        .catch((error: unknown) => {
+          // Should not block if interaction is not found
+          if (error instanceof errors.SessionNotFound) {
+            return null;
```
should not guard the session strictly? If no session is found return just return the default language settings.

simeng-li commented on packages/core/src/routes/phrase.ts:
```diff
@@ -0,0 +1,130 @@
+const isBuiltInLanguageKey = (key: string): key is LanguageKey =>
+  Object.keys(resource).includes(key) && languageKeyGuard.safeParse(key).success;
+
+const isFullyTranslated = (fullTranslation: Translation, targetTranslation: Translation) => {
```
deep merge will do all the works, deep merge(custom, build in target, en custom, build in en);

simeng-li commented on packages/core/src/routes/phrase.ts:
```diff
@@ -0,0 +1,130 @@
+  const hasCustomPhrase = customLanguageKeys.includes(languageKey);
+  const hasBuiltInPhrase = isBuiltInLanguageKey(languageKey);
+
+  if (!hasCustomPhrase && !hasBuiltInPhrase) {
+    return getPhrase(defaultLanguageKey);
+  }
+
+  if (!hasCustomPhrase && hasBuiltInPhrase) {
+    return { languageKey, ...resource[languageKey] };
+  }
+
+  const customPhrase = await findCustomPhraseByLanguageKey(languageKey);
+
+  if (isFullyTranslated(en.translation, customPhrase.translation)) {
+    return customPhrase;
+  }
+
+  if (hasBuiltInPhrase) {
+    return deepmerge(resource[languageKey], customPhrase);
+  }
+
+  const defaultPhrase = await getPhrase(defaultLanguageKey);
+
+  return deepmerge(defaultPhrase, customPhrase);
```
```js

const customPhrase = await findCustomPhraseByLanguageKey(languageKey);
const buidInPhrase = resource['languageKey'];

if (buidInPhrase) {
  return deepMerge(customPhrase, buidInPhrase);
}


const customFallback = await findCustomPhraseByLanguageKey(fallback);
const buildInFallback = resource [fallback];

if (buildInFallback) {
  return deepMerge(customPhrase, customFallback, buildInFallback );
}

const customEn = await findCustomPhraseByLanguageKey('en');
const buildInEn = resource.en;

return deepMerge( customPhrase, customFallback, customEn, buildInEn);

```

IceHe commented on packages/core/src/routes/phrase.ts:
```diff
@@ -0,0 +1,130 @@
+
+import { AnonymousRouter } from './types';
+
+const getLanguageInfo = async (applicationId: unknown) => {
+  if (applicationId === adminConsoleApplicationId) {
+    return adminConsoleSignInExperience.languageInfo;
+  }
```
LOG-4212

xiaoyijun commented on packages/core/src/routes/phrase.ts:
```diff
@@ -0,0 +1,107 @@
+};
+
+const isBuiltInLanguageKey = (key: string): key is LanguageKey =>
+  Object.keys(resource).includes(key) && languageKeyGuard.safeParse(key).success;
```
`languageKeyGuard.safeParse(key).success` this code seems redundant and maybe cause errors in the future.
Because our built-in language comes from `resource`, and the `languageKey` represent the supported language keys (including custom languages).

wangsijie commented on packages/core/src/routes/phrase.ts:
```diff
@@ -0,0 +1,107 @@
+        .interactionDetails(ctx.req, ctx.res)
+        .catch((error: unknown) => {
+          // Should not block when failed to get interaction
+          return null;
```
Can we make this block "empty"?

xiaoyijun commented on packages/core/src/lib/phrase.ts:
```diff
@@ -0,0 +1,46 @@
+
+const has = (object: Record<string, unknown>) => Object.keys(object).length > 0;
+
+export const getPhrase = async (detectedLanguages: string[], fallbackLanguage: string) => {
```
Maybe we can set the `fallbackLanguage` type as `LanguageKey`.
Because the fully supported language keys haven't been determined, we can add an issue to track this.

IceHe replied:
Intend to do it later. Added LOG-4221 to track it.

IceHe replied:
Intend to do it later. Added LOG-4221 to track it.

darcyYe commented on packages/core/src/lib/phrase.ts:
```diff
@@ -0,0 +1,46 @@
+const getBuiltInPhrase = (key: string | undefined) =>
+  key && isBuiltInLanguage(key) ? { ...resource[key], languageKey: key } : {};
+
+const has = (object: Record<string, unknown>) => Object.keys(object).length > 0;
```
nit:
`has` seems unclear? maybe using `isNotEmpty()` or something else?
And as this is only used in `getPhrase()`, we may simply defined this method at the beginning of `getPhrase()`?

IceHe replied:
refactored and this function is unneeded.

darcyYe reviewed (approved):
Overall LGTM

simeng-li commented on packages/core/src/lib/phrase.ts:
```diff
@@ -0,0 +1,46 @@
+
+import { findAllCustomLanguageKeys, findCustomPhraseByLanguageKey } from '@/queries/custom-phrase';
+
+const isBuiltInLanguage = (key: string): key is LanguageKey => Object.keys(resource).includes(key);
```
```suggestion
const isBuiltInLanguage = (key: string): key is LanguageKey => Object.keys(resource).includes(key);

const getCustomPhrases = async (key: string | undefined, customLanguages: string[]) =>
  key && customLanguages.includes(key) ? findCustomPhraseByLanguageKey(key) : undefined;

const getBuiltInPhrases = (key: string | undefined) =>
  key && isBuiltInLanguage(key) ? resource[key] : undefined;

type PhrasesPool = {
  customPhrases?: CustomPhrase;
  buildInPhrases?: ResourceLanguage;
};

const getLanguagePhrases = async (
  key: string | undefined,
  customLanguages: string[]
): Promise<PhrasesPool | undefined> => {
  const customPhrases = await getCustomPhrases(key, customLanguages);
  const buildInPhrases = getBuiltInPhrases(key);

  if (!customPhrases && !buildInPhrases) {
    return;
  }

  return {
    customPhrases,
    buildInPhrases,
  };
};

const getFallbackLanguage = async (fallbackLanguage: string, customLanguages: string[]) => {
  const fallbackLanguagePhrases = await getLanguagePhrases(fallbackLanguage, customLanguages);

  if (!fallbackLanguagePhrases) {
    throw new Error("Fallback Language Phrases won't be empty");
  }

  return fallbackLanguagePhrases;
};

const getEnLanguage = async (customLanguages: string[]) => {
  const enLanguagePhrase = await getLanguagePhrases('en', customLanguages);

  if (!enLanguagePhrase) {
    throw new Error("EN language Phrases won't be empty");
  }

  if (!enLanguagePhrase.buildInPhrases) {
    throw new Error("EN build In language Phrases won't be empty");
  }

  return {
    customPhrases: enLanguagePhrase.customPhrases,
    buildInPhrases: enLanguagePhrase.buildInPhrases,
  };
};

export const getPhrase = async (
  detectedLanguages: string[],
  fallbackLanguage: LanguageKey
): Promise<ResourceLanguage & { key: string }> => {
  const customLanguages = await findAllCustomLanguageKeys();

  const detectLanguage = detectedLanguages.find(
    (key) => isBuiltInLanguage(key) || customLanguages.includes(key)
  );

  const detectedLanguagePhrases = await getLanguagePhrases(detectLanguage, customLanguages);

  // Detect Custom x Detect BuildIn
  if (detectLanguage && detectedLanguagePhrases?.buildInPhrases) {
    return {
      ...deepmerge(
        detectedLanguagePhrases.customPhrases ?? {},
        detectedLanguagePhrases.buildInPhrases
      ),
      key: detectLanguage,
    };
  }

  const fallbackLanguagePhrases = await getFallbackLanguage(fallbackLanguage, customLanguages);

  // Detect Custom x Fallback Custom x Fallback BuildIn
  if (fallbackLanguagePhrases.buildInPhrases) {
    return {
      ...deepmerge.all<ResourceLanguage>([
        detectedLanguagePhrases?.customPhrases ?? {},
        fallbackLanguagePhrases.customPhrases ?? {},
        fallbackLanguagePhrases.buildInPhrases,
      ]),
      key: detectLanguage ?? fallbackLanguage,
    };
  }

  const enLanguagePhrase = await getEnLanguage(customLanguages);

  // Detect Custom x Fallback Custom x EN Custom x EN BuildIn
  return {
    ...deepmerge.all<ResourceLanguage>([
      detectedLanguagePhrases?.customPhrases ?? {},
      fallbackLanguagePhrases.customPhrases ?? {},
      enLanguagePhrase.customPhrases ?? {},
      enLanguagePhrase.buildInPhrases,
    ]),
    key: detectLanguage ?? fallbackLanguage,
  };
};

IceHe replied:
refactored. do it a simpler way.

wangsijie reviewed (approved):
Overall LGTM, it would be better if functions in the "route" file are tested in seperate, treat them as "units"

IceHe commented:
Updated
- @wangsijie Extract `getPhrase` to `lib/phrase` and test separately.
- @simeng-li Add `type Phrase = typeof en` for accurate type check
- Add `fullTranslatedGuard` as `GET /phrase` response guard to ensure the final translation is  fully-translated or fails fast

simeng-li reviewed (approved):
LGTM

IceHe commented:
Updated:
- Rename to accurate type name `type UiPhrase = typeof en`.
- Remove `fullTranslatedGuard` in the `GET /phrase response guard` to avoid too many checks.
- Use `cleanDeep` on the custom phrase before merging with the built-in phrase to avoid empty string values in the final phrase.

wangsijie reviewed (approved):
LGTM

IceHe commented:
Rebased the feature branch [feature-manage-language](https://github.com/logto-io/logto/tree/feature-manage-language) and changed target branch to [feature-manage-language](https://github.com/logto-io/logto/tree/feature-manage-language)

xiaoyijun commented on packages/phrases-ui/src/types.ts:
```diff
@@ -2,11 +2,8 @@ import { LanguageKey, languageKeyGuard } from '@logto/core-kit';
-
-/* Copied from i18next/index.d.ts */
-export type Resource = Record<LanguageKey, LocalePhrase>;
+export type UiPhrase = typeof en;
+export type UiResource = Record<LanguageKey, UiPhrase>;
```
how about:
```ts
export type LocalePhrase = typeof en;
export type Resource = Record<LanguageKey, LocalePhrase>;
```

xiaoyijun reviewed (approved):
LGTM
